Generate serialization of WCBackingStore

donny-dont opened the pull request:
#### e45afd87e47f85c70bd314c8a22de09f2c31e010
<pre>
Generate serialization of WCBackingStore
<a href="https://bugs.webkit.org/show_bug.cgi?id=268502">https://bugs.webkit.org/show_bug.cgi?id=268502</a>

Reviewed by Alex Christensen.

Modify `WCBackingStore` so its serialization can be generated. Pass an
`ImageBufferBackendHandle` through IPC. Add a constructor to create from
a handle, and a method to retrieve a handle from an instance. Adding
these allows the serialization framework to work on the class.

* Source/WebKit/PlatformWin.cmake:
* Source/WebKit/WebProcess/WebPage/wc/WCBackingStore.cpp: Copied from Source/WebKit/WebProcess/WebPage/wc/WCBackingStore.h.
(WebKit::WCBackingStore::WCBackingStore):
(WebKit::WCBackingStore::handle const):
* Source/WebKit/WebProcess/WebPage/wc/WCBackingStore.h:
(WebKit::WCBackingStore::encode const): Deleted.
(WebKit::WCBackingStore::decode): Deleted.
* Source/WebKit/WebProcess/WebPage/wc/WCBackingStore.serialization.in: Added.

Canonical link: <a href="https://commits.webkit.org/273938@main">https://commits.webkit.org/273938@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a70d8f76999eed3b7719c054c8f1f39fdc18c8d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37245 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16147 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39542 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39808 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/33218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18727 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13289 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37808 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13614 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32758 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11865 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11841 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41068 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33718 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/33837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12192 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9976 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/35906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8412 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12543 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/WebPage/wc/WCBackingStore.cpp:
```diff
@@ -0,0 +1,55 @@
+    if (!is<ImageBufferBackendHandleSharing>(sharing))
+        return std::nullopt;
+
+    return downcast<ImageBufferBackendHandleSharing>(*sharing).createBackendHandle();
```
Could use dynamicDowncast instead of is then downcast.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/wc/WCBackingStore.cpp:
```diff
@@ -0,0 +1,55 @@
+WCBackingStore::WCBackingStore(std::optional<ImageBufferBackendHandle>&& handle)
+{
+    if (handle)
+        m_bitmap = ShareableBitmap::create(WTFMove(std::get<ShareableBitmap::Handle>(*handle)));
```
You probably also want a std::get_if

donny-dont committed:
Generate serialization of WCBackingStore
https://bugs.webkit.org/show_bug.cgi?id=268502

Reviewed by Alex Christensen.

Modify `WCBackingStore` so its serialization can be generated. Pass an
`ImageBufferBackendHandle` through IPC. Add a constructor to create from
a handle, and a method to retrieve a handle from an instance. Adding
these allows the serialization framework to work on the class.

* Source/WebKit/PlatformWin.cmake:
* Source/WebKit/WebProcess/WebPage/wc/WCBackingStore.cpp: Copied from Source/WebKit/WebProcess/WebPage/wc/WCBackingStore.h.
(WebKit::WCBackingStore::WCBackingStore):
(WebKit::WCBackingStore::handle const):
* Source/WebKit/WebProcess/WebPage/wc/WCBackingStore.h:
(WebKit::WCBackingStore::encode const): Deleted.
(WebKit::WCBackingStore::decode): Deleted.
* Source/WebKit/WebProcess/WebPage/wc/WCBackingStore.serialization.in: Added.

Canonical link: https://commits.webkit.org/273938@main
